src/ui/groups_widgets.cc

```diff
@@ -772,6 +773,57 @@ void DictGroupsWidget::addAutoGroups()
   }
 }
 
+
+void DictGroupsWidget::addAutoGroupsByFolders()
+{
+  if ( activeDicts->empty() ) {
+    return;
+  }
+
+  if ( QMessageBox::information( this,
+                                 tr( "Confirmation" ),
+                                 tr( "Are you sure you want to generate a set of groups "
+                                     "based on containing folders?" ),
+                                 QMessageBox::Yes,
+                                 QMessageBox::Cancel )
+       != QMessageBox::Yes ) {
+    return;
+  }
+
+  QMultiMap< QString, sptr< Dictionary::Class > > dictMap;
+
+  for ( const auto & dict : *activeDicts ) {
+    QString dictFolder = dict->getContainingFolder();
+
+    if ( dictFolder.isEmpty() ) {
+      continue;
+    }
+
+    QDir dir = dictFolder;
+    if ( dir.cdUp() ) {
+      dictMap.insert( dir.dirName(), dict );
+    }
+    else {
+      qWarning() << "Cannot auto group because parent folder cannot be reached: " << dir;
+      continue;
+    }
+  }
```

Comment:
This `dir` is obtained from `Dictioanry::Class::getContainingFolder` which read the first file from a list of `dictionaryFiles` thus it will be guaranteed to exist.